src/sentry/api/endpoints/user_role_details.py

```diff
@@ -0,0 +1,77 @@
+import logging
+
+from django.db import IntegrityError, transaction
+from rest_framework.request import Request
+from rest_framework.response import Response
+
+from sentry.api.bases.user import UserEndpoint
+from sentry.api.decorators import sudo_required
+from sentry.api.permissions import SuperuserPermission
+from sentry.models import UserRole
+
+audit_logger = logging.getLogger("sentry.audit.user")
+
+
+class UserUserRoleDetailsEndpoint(UserEndpoint):
+    permission_classes = (SuperuserPermission,)
+
+    def get(self, request: Request, user, role_name) -> Response:
+        # XXX(dcramer): we may decide to relax "view" permission over time, but being more restrictive by default
+        # is preferred
+        if not request.access.has_permission("users.admin"):
+            return self.respond(status=403)
+
+        has_perm = UserRole.objects.filter(user=user, permission=role_name).exists()
+        return self.respond(status=204 if has_perm else 404)
+
+    @sudo_required
+    def post(self, request: Request, user, role_name) -> Response:
+        if not request.access.has_permission("users.admin"):
+            return self.respond(status=403)
+
+        try:
+            role = UserRole.objects.get(user=user, name=role_name)
+        except UserRole.DoesNotExist:
+            return self.respond({"detail": f"'{role_name}' is not a known role."}, status=404)
+
+        try:
+            with transaction.atomic():
+                role.users.add(user)
+                audit_logger.info(
+                    "user.add-role",
+                    extra={
+                        "actor_id": request.user.id,
+                        "user_id": user.id,
+                        "role_id": role.id,
+                    },
+                )
+        except IntegrityError as e:
+            if "already exists" in str(e):
+                return self.respond(status=410)
+            raise
+        return self.respond(status=201)
+
+    @sudo_required
+    def delete(self, request: Request, user, role_name) -> Response:
+        if not request.access.has_permission("users.admin"):
+            return self.respond(status=403)
+
+        try:
+            role = UserRole.objects.get(user=user, name=role_name)
+        except UserRole.DoesNotExist:
+            return self.respond({"detail": f"'{role_name}' is not a known role."}, status=404)
+
+        with transaction.atomic():
+            with transaction.atomic():
```

Comment:
are the double `transaction.atomic()`'s intentional here?

Reply:
nope good catch! will fix monday